File: day_2/1.py
sum = 0

with open("2.dat", "r") as f:
    for ranges in f.readlines()[0].split(","):
        start, end = ranges.split("-")
        for number in range(int(start), int(end) + 1):
            length = len(str(number))
            if str(number)[:length//2] == str(number)[length//2:]:
                sum += number
    print(sum)
            

File: day_2/1.hs
main :: IO ()
main = do
    contents <- readFile "2.dat"
    let firstLine = takeWhile (/= '\n') contents
        ranges    = splitBy ',' firstLine
        nums      = concatMap expandRange ranges
        goodNums  = filter halfEqual nums
    print (sum goodNums)

splitBy :: Char -> String -> [String]
splitBy _ "" = [""]
splitBy d (c:cs)
    | c == d    = "" : rest
    | otherwise = (c : head rest) : tail rest
  where
    rest = splitBy d cs

expandRange :: String -> [Int]
expandRange s =
    let [a, b] = splitBy '-' s
    in [read a .. read b]

halfEqual :: Int -> Bool
halfEqual n =
    let s = show n
        len = length s
        (l, r) = splitAt (len `div` 2) s
    in even len && l == r
